driver-core/src/main/com/mongodb/internal/Timeout.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.internal;
+
+import com.mongodb.annotations.Immutable;
+
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+
+import static com.mongodb.assertions.Assertions.notNull;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+/**
+ * A <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/doc-files/ValueBased.html">value-based</a> class
+ * useful for tracking timeouts.
+ */
+@Immutable
+public final class Timeout {
+    private static final Timeout INFINITE = new Timeout(-1, 0);
+    private static final Timeout IMMEDIATE = new Timeout(0, 0);
+
+    private final long durationNanos;
+    private final long startNanos;
+
+    private Timeout(final long durationNanos, final long startNanos) {
+        this.durationNanos = durationNanos;
+        this.startNanos = startNanos;
+    }
+
+    /**
+     * @see #startNow(long)
+     */
+    public static Timeout startNow(final long duration, final TimeUnit unit) {
+        return startNow(notNull("unit", unit).toNanos(duration));
+    }
+
+    /**
+     * Returns an {@linkplain #isInfinite() infinite} timeout if {@code durationNanos} is either negative
+     * or is equal to {@link Long#MAX_VALUE},
+     * an {@linkplain #isImmediate() immediate} timeout if {@code durationNanos} is 0,
+     * otherwise an object that represents the specified {@code durationNanos}.
+     */
+    public static Timeout startNow(final long durationNanos) {
+        if (durationNanos < 0 || durationNanos == Long.MAX_VALUE) {
+            return INFINITE;
+        } else if (durationNanos == 0) {
+            return IMMEDIATE;
+        } else {
+            return new Timeout(durationNanos, System.nanoTime());
+        }
+    }
+
+    /**
+     * @see #startNow(long)
```

Comment:
I'm not sure of the value of doing this javadoc redirection in internal code.

Reply:
It allows a documentation reader to jump to the place that has the relevant documentation. Whether this reader is internal (you, me) or external is not relevant.

Reply:
Ack